File: leetcode/05-String-to-Integer/python/main.py
class Solution:
    def myAtoi(self, s: str) -> int:
        result = 0

        # 変換表
        map = {
            '1': 1,
            '2': 2,
            '3': 3,
            '4': 4,
            '5': 5,
            '6': 6,
            '7': 7,
            '8': 8,
            '9': 9,
            '0': 0,
        }

        # 頭の空白を取り除く
        for i in range(len(s)):
            if not s[i] == ' ':
                s = s[i:]
                break

        # 符号を確認する
        negative = False
        if len(s) > 0 and s[0] == '-':
            # 先頭にマイナスがある時
            negative = True
            s = s[1:]
        elif len(s) > 0 and s[0] == '+':
            # 先頭にプラスがある時
            negative = False
            s = s[1:]
        # 先頭に符号がない時は何もしない

        # 数字以外が現れるまで読み込む
        for i in range(len(s)):
            if not s[i] in map:
                s = s[:i]
                break

        # 数字に変換する
        for i in range(len(s)):
            currentCar = s[i]
            currentNum = map[s[i]]

            result += currentNum * 10 ** (len(s) - 1 - i)

        # 符号を反映する
        result = result * -1 if negative else result

        # 範囲を確認
        result = (-2)**31 if result < (-2)**31 else result
        result = 2**31-1 if 2**31-1 < result else result

        return result

File: leetcode/05-String-to-Integer/go/main.go
package main

import (
	"fmt"
	"math"
	"strconv"
	"strings"
	"unicode"
)

func _myAtoi(s string) int {
	sign := 1
	res := 0

	s = strings.TrimLeft(s, " ")

	if len(s) == 0 {
		return res
	}

	if s[0] == '-' {
		sign = -1
		s = s[1:]
	} else if s[0] == '+' {
		s = s[1:]
	}

	for _, ch := range s {
		if !unicode.IsDigit(ch) {
			break
		}

		dig := int(ch - '0')
		res = res*10 + dig

		if sign*res < math.MinInt32 {
			return math.MinInt32
		} else if sign*res > math.MaxInt32 {
			return math.MaxInt32
		}
	}

	return sign * res
}

func myAtoi(s string) int {
	result := 0
	numString := ""

	isLeadingWhitespace := true
	signChar := ""

	numMap := make(map[string]int)
	for i := 0; i < 10; i++ {
		numMap[strconv.Itoa(i)] = i
	}

	for i := 0; i < len(s); i++ {
		currentChar := s[i : i+1]

		// 先頭のスペースは読み飛ばす
		if isLeadingWhitespace && currentChar == " " {
			continue
		}
		if currentChar != " " {
			isLeadingWhitespace = false
		}

		// 符号が出てきたら格納する
		if numString == "" && signChar == "" && currentChar == "-" {
			signChar = "-"
			continue
		}
		if numString == "" && signChar == "" && currentChar == "+" {
			signChar = "+"
			continue
		}

		// 数字以外が出てきたらループを抜ける
		if _, isNum := numMap[currentChar]; !isNum {
			break
		}

		// 数字であればそれを控える
		numString += currentChar
	}
	fmt.Printf("numString: %v\n", numString)

	// 数字に変換する
	var tmp float64
	for i := 0; i < len(numString); i++ {
		tmp += float64(numMap[numString[i:i+1]]) * math.Pow10(len(numString)-i-1)
	}
	fmt.Printf("tmp: %v\n", tmp)

	// 符号を適用する
	if signChar == "-" {
		tmp = tmp * -1
	}

	// 範囲を確認する
	if tmp < math.Pow(-2, 31) {
		result = int(math.Pow(-2, 31))
	} else if tmp > math.Pow(2, 31)-1 {
		result = int(math.Pow(2, 31)) - 1
	} else {
		result = int(tmp)
	}
	fmt.Printf("result: %v\n", result)

	return result
}
